vulkan_generator/handle_remapper/generator.py

```diff
@@ -0,0 +1,324 @@
+            if(map_count_iter == {map_count_name}.end() || map_count_iter->second <= 0) throw InternalConsistencyException();
+            if(map_iter->second != replayHandle) throw NonDispatchableHandleRedefinitionException();
+        }}
+
+        {map_name}[captureHandle] = replayHandle;
+        {map_count_name}[captureHandle] = {map_count_name}[captureHandle] +1;"""
+    )
+
+def non_dispatchable_handle_remove_code(handle: str) -> str:
+    map_name = handle_map_name(handle)
+    map_count_name = handle_count_map_name(handle)
+    return dedent(f"""
+        auto map_iter = {map_name}.find(captureHandle);
+        auto map_count_iter = {map_count_name}.find(captureHandle);
+
+        if(map_iter == {map_name}.end() || map_count_iter == {map_count_name}.end()) {{
+            throw RemoveNonExistantHandleException();
+        }}
+        else {{
+            if(map_count_iter->second <= 0) throw InternalConsistencyException();
+        }}
+
+        if(({map_count_name}[captureHandle] = {map_count_name}[captureHandle] -1) <= 0){{
+            {map_name}.erase(map_iter);
+            {map_count_name}.erase(map_count_iter);
+        }}"""
+    )
+
+def non_dispatchable_handle_remap_code(handle: str) -> str:
+    map_name = handle_map_name(handle)
+    map_count_name = handle_count_map_name(handle)
+    return dedent(f"""
+        auto map_iter = {map_name}.find(captureHandle);
+        auto map_count_iter = {map_count_name}.find(captureHandle);
+
+        if(map_iter == {map_name}.end() || map_count_iter == {map_count_name}.end()) {{
+            throw RemapNonExistantHandleException();
+        }}
+        else {{
+            if(map_count_iter->second <= 0) throw InternalConsistencyException();
+        }}
+
+        return map_iter->second;"""
+    )
+
+def handle_add_code(handle: str, all_vulkan_types : types.AllVulkanTypes) -> str:
+    if all_vulkan_types.handles[handle].dispatchable:
+        return dispatchable_handle_add_code(handle)
+    else:
+        return non_dispatchable_handle_add_code(handle)
+
+def handle_remove_code(handle: str, all_vulkan_types : types.AllVulkanTypes) -> str:
+    if all_vulkan_types.handles[handle].dispatchable:
+        return dispatchable_handle_remove_code(handle)
+    else:
+        return non_dispatchable_handle_remove_code(handle)
+
+def handle_remap_code(handle: str, all_vulkan_types : types.AllVulkanTypes) -> str:
+    if all_vulkan_types.handles[handle].dispatchable:
+        return dispatchable_handle_remap_code(handle)
+    else:
+        return non_dispatchable_handle_remap_code(handle)
+
+def generate_handle_remapper_h(file_path : Path, all_vulkan_types : types.AllVulkanTypes) :
+
+    with open(file_path, "w", encoding="ascii") as remapper_h:
+
+        remapper_h.write(codegen.generated_header())
+
+        remapper_h.write(dedent("""
+            #include <map>
+            #include <iostream>
+            #include <map>
+
+            namespace agi {
+            namespace replay2 {
```

Comment:
Just name it "replay". The current replayer is in the "gapir" namespace, so as long as it's in an "agi" namespace, there is no collision.

Reply:
A problem we have with the current AGI codebase is the overloading of many common terms like builder, data, replay, etc. When talking about AGI it's often not easy to understand which "builder" or "replay" someone is referring to. Of course this could be solved by referring to everything in speech with it's full package/namespace scope like "gapir/replay" or "agi/replay" but nobody really speaks like this. I'm not super hung up on the name replay2 but it would be nice to give it a name that could be spoken about in English without prefacing namespaces.